compiler/rustc_abi/src/layout.rs

```diff
@@ -934,3 +775,287 @@ pub trait LayoutCalculator {
+            }
+            // Otherwise we just leave things alone and actually optimize the type's fields
+        } else {
+            let max_field_align = fields.iter().map(|f| f.align().abi.bytes()).max().unwrap_or(1);
+            let any_niche = fields.iter().any(|f| f.largest_niche().is_some());
+            let effective_field_align = |layout: Layout<'_>| {
+                if let Some(pack) = pack {
+                    // return the packed alignment in bytes
+                    layout.align().abi.min(pack).bytes()
+                } else {
+                    // returns log2(effective-align).
+                    // This is ok since `pack` applies to all fields equally.
+                    // The calculation assumes that size is an integer multiple of align, except for ZSTs.
+                    //
+                    // group [u8; 4] with align-4 or [u8; 6] with align-2 fields
+                    let align = layout.align().abi.bytes();
+                    let size = layout.size().bytes();
+                    let size_as_align = align.max(size).trailing_zeros();
+                    let size_as_align = if any_niche {
+                        max_field_align.trailing_zeros().min(size_as_align)
+                    } else {
+                        size_as_align
+                    };
+                    size_as_align as u64
+                }
+            };
+
+            match kind {
+                StructKind::AlwaysSized | StructKind::MaybeUnsized => {
+                    optimizing.sort_by_key(|&x| {
+                        let f = fields[x];
+                        let field_size = f.size().bytes();
+                        let niche_size = f.largest_niche().map_or(0, |n| n.available(dl));
+                        let niche_size = if niche_bias_start {
+                            u128::MAX - niche_size // large niche first
+                        } else {
+                            niche_size // large niche last
+                        };
+                        let inner_niche_placement = if niche_bias_start {
+                            f.largest_niche().map_or(0, |n| n.offset.bytes())
+                        } else {
+                            f.largest_niche().map_or(0, |n| {
+                                field_size - n.value.size(dl).bytes() - n.offset.bytes()
+                            })
+                        };
+
+                        // Place ZSTs first to avoid "interesting offsets", especially with only one
+                        // or two non-ZST fields. This helps Scalar/ScalarPair layouts.
+                        // Then place largest alignments first.
+                        // Then prioritize niche placement within alignment group according to
+                        // `niche_bias_start`.
+                        // Then among fields with equally-sized niches prefer the ones
+                        // closer to the start/end of the field.
```

Comment:
Why does this matter?

Reply:
Let's say we have

```
#[repr(C)]
struct A(u8, bool, u8);

struct B(bool);

struct C(A, B);
```

Then we have two niche-candidates (both bools). Since we try to get the niches close to the edge of a struct we try to prioritize B in this case because its niche sits closer to the start/end of the field.

Reply:
But why do we want niches close to the start/end?

Reply:
dtolnay has a good explanation in #104807

If the niche is at the edge then it's more likely that the niche-as-discriminant encoding will end up being beneficial. Otherwise it'll pick the explicit-discriminant layout which results in larger enums. Or at least it'll have to arrange the variants in a less compact way.